### ISSUE TITLE
Cannot add item to PublishQueue

### ISSUE THREAD

**Issue by petrformanek** [sergeyshushlyapin/Sitecore.FakeDb#146]
FakeDb doesn't implement working with PublishQueue.
Calling PublishManager.AddToPublishQueue do just nothing (FakeDb provides just empty implementation). The same is for PublishManager.GetPublishQueue.

Can you implement working with PublishQueue?
I was trying to implement my own DataProvider which inherited from FakeDataProvider but I was unable to hook it to sitecore (datProviders sections in App.config is ignored and FakeDataProvider is always used).

Found in version 1.2.2.

Thanks
Petr

**Comment by pveller** [comment]
I think you will hit a larger wall with testing your own Data Provider with FakeDb until we migrate to command-less approach to faking. Some of it is explained in #138 (at the bottom of the thread). FakeDb right now intercepts at the database command level so not many things will reach the data provider chain as you would normally expect in a the real Sitecore instance. @sergeyshushlyapin has been working on a Data Provider based implementation that will likely play nicely with the chaining that you would expect out of the box but you will have to be careful not to chain yourself in front of FakeDB's one. 

That said, can you please provide more details about your publishing queue scenario? What exactly are you trying to test with FakeDb that requires the queue working?

**Comment by petrformanek** [comment]
We have a quite large sitecore deployment and have one let's say master sitecore instance and several regional sitecore instances. On master instance we regularly read changed items from PublishQueue and we push these items to regional instances. If push fails then item is updated in PublishQueue to re-try it in the next round.
We call just PublishManager.AddToPublishQueue and PublishManager.GetPublishQueue.

I was able to make it work with following hack:

```C#
_db = new Db();
_sourceDb = _db.Database;

var dataProvidersField = _sourceDb.GetType().GetField("_dataProviders", BindingFlags.Instance | BindingFlags.NonPublic);
var dataProvidersCollection = (DataProviderCollection)dataProvidersField.GetValue(_sourceDb);
dataProvidersCollection[0] = new OriFakeDataProvider(((FakeDataProvider)dataProvidersCollection[0]).DataStorage);

public class OriFakeDataProvider : FakeDataProvider
{
	readonly HashSet<ID> _publishingQueue = new HashSet<ID>();

	public OriFakeDataProvider() { }

	public OriFakeDataProvider(DataStorage dataStorage) : base(dataStorage) { }

	public override bool AddToPublishQueue(ID itemID, string action, DateTime date, string language, CallContext context)
	{
		_publishingQueue.Add(itemID);
		return true;
	}

	public override IDList GetPublishQueue(DateTime from, DateTime to, CallContext context)
	{
		return IDList.Build(_publishingQueue.ToArray());
	}
}

```

**Comment by pveller** [comment]
Oh, I see now. So you really just need the queue methods to allow to `add` and `get`. That's easy :) I am sure it will soon be on the NuGet feed. Stay tuned!

**Comment by petrformanek** [comment]
Great! 
If it would also respect date from and to it would be great.

**Comment by sergeyshushlyapin** [comment]
Is `HashSet` suitable for that purpose? Shouldn't it be possible to add the same *itemId* multiple times?

**Comment by petrformanek** [comment]
HashSet is not good for implementation. Sitecore can store same itemId to PublishingQueue

**Comment by sergeyshushlyapin** [comment]
Now available as [v1.3.0](https://www.nuget.org/packages/Sitecore.FakeDb/).

**Comment by petrformanek** [comment]
Thanks for an update. I tried v1.3.0 today and there is one small bug in GetPublishQueue method. When the PublishQueue is empty (AddToPublishQueue was not called yet) it throws 

System.ArgumentNullException: Value cannot be null.
Parameter name: source
    at System.Linq.Enumerable.Where[TSource](IEnumerable`1 source, Func`2 predicate)
   at Sitecore.FakeDb.Data.DataProviders.FakeDataProvider.GetPublishQueue(DateTime from, DateTime to, CallContext context)

this.publishQueue.Value is not initialized yet.

Please fix it.
Thanks

**Comment by sergeyshushlyapin** [comment]
Thanks @petrformanek. Fixed in [v1.3.1](https://www.nuget.org/packages/Sitecore.FakeDb/1.3.1).
